runtime/src/storage/iouring.rs

```diff
@@ -0,0 +1,450 @@
+}
+
+fn new_ring(cfg: &IoUringConfig) -> Result<IoUring, std::io::Error> {
+    let mut builder = &mut IoUring::builder();
+    if cfg.iopoll {
+        builder = builder.setup_iopoll();
+    }
+    if cfg.single_issuer {
+        builder = builder.setup_single_issuer();
+    }
+    builder.build(cfg.size)
+}
+
+/// Background task that polls for completed work and notifies waiters on completion.
+/// The user data field of all operations received on `receiver` will be ignored.
+async fn do_work(
+    cfg: IoUringConfig,
+    mut receiver: mpsc::Receiver<(SqueueEntry, oneshot::Sender<i32>)>,
+) {
+    let mut ring = new_ring(&cfg).expect("unable to create io_uring instance");
+    let mut next_work_id: u64 = 0;
+    // Maps a work ID to the sender that we will send the result to.
+    let mut waiters: HashMap<_, oneshot::Sender<i32>> = HashMap::with_capacity(cfg.size as usize);
+
+    loop {
+        // Try to get a completion
+        if let Some(cqe) = ring.completion().next() {
+            let work_id = cqe.user_data();
+            let result = cqe.result();
+            let sender = waiters.remove(&work_id).expect("work is missing");
+            // Notify with the result of this operation
+            let _ = sender.send(result);
+            continue;
+        }
+
+        // Try to fill the submission queue with incoming work.
+        // Stop if we are at the max number of processing work.
+        while waiters.len() < cfg.size as usize {
+            // Wait for more work
+            let (mut work, sender) = if waiters.is_empty() {
+                // Block until there is something to do
+                match receiver.next().await {
+                    Some(work) => work,
+                    None => return,
+                }
+            } else {
+                // Handle incoming work
+                match receiver.try_next() {
+                    // Got work without blocking
+                    Ok(Some(work_item)) => work_item,
+                    // Channel closed, shut down
+                    Ok(None) => return,
+                    // No new work available, wait for a completion
+                    Err(_) => break,
+                }
+            };
+
+            // Assign a unique id
+            let work_id = next_work_id;
+            work = work.user_data(work_id);
+            // Use wrapping add in case we overflow
+            next_work_id = next_work_id.wrapping_add(1);
+
+            // We'll send the result of this operation to `sender`.
+            waiters.insert(work_id, sender);
+
+            // Submit the operation to the ring
+            unsafe {
+                ring.submission()
+                    .push(&work)
+                    .expect("unable to push to queue");
+            }
+        }
+
+        // Note that waiters.len() is never empty here.
+        // If it were, this would block forever.
+        ring.submit_and_wait(1).expect("unable to submit to ring");
```

Comment:
May be worth nothing that the argument for `wait` here means wait until there is at least 1 item in the completion queue (not that 1 new item must accrue, which may have arrived after we last checked the completion queue). This would otherwise make you think a deadlock was possible.

https://man7.org/linux/man-pages/man2/io_uring_enter.2.html

Reply:
updated. Lmk if the new comment looks OK to you.